pandas/tests/generic/methods/test_sample.py

```diff
@@ -0,0 +1,309 @@
+        with pytest.raises(ValueError):
+            obj.sample(n=-3)
+        with pytest.raises(ValueError):
+            obj.sample(frac=-0.3)
+
+    def test_sample_requires_integer_n(self, obj):
+        # Make sure float values of `n` give error
+        with pytest.raises(ValueError):
+            obj.sample(n=3.2)
+
+    def test_sample_invalid_weight_lengths(self, obj):
+        # Weight length must be right
+        with pytest.raises(ValueError):
+            obj.sample(n=3, weights=[0, 1])
+
+        with pytest.raises(ValueError):
+            bad_weights = [0.5] * 11
+            obj.sample(n=3, weights=bad_weights)
+
+        with pytest.raises(ValueError):
+            bad_weight_series = Series([0, 0, 0.2])
+            obj.sample(n=4, weights=bad_weight_series)
+
+    def test_sample_negative_weights(self, obj):
+        # Check won't accept negative weights
+        with pytest.raises(ValueError):
+            bad_weights = [-0.1] * 10
+            obj.sample(n=3, weights=bad_weights)
+
+    def test_sample_inf_weights(self, obj):
+        # Check inf and -inf throw errors:
+
+        with pytest.raises(ValueError):
+            weights_with_inf = [0.1] * 10
+            weights_with_inf[0] = np.inf
+            obj.sample(n=3, weights=weights_with_inf)
+
+        with pytest.raises(ValueError):
+            weights_with_ninf = [0.1] * 10
+            weights_with_ninf[0] = -np.inf
+            obj.sample(n=3, weights=weights_with_ninf)
+
+    def test_sample_zero_weights(self, obj):
+        # All zeros raises errors
+
+        zero_weights = [0] * 10
+        with pytest.raises(ValueError):
+            obj.sample(n=3, weights=zero_weights)
+
+    def test_sample_missing_weights(self, obj):
+        # All missing weights
+
+        nan_weights = [np.nan] * 10
+        with pytest.raises(ValueError):
+            obj.sample(n=3, weights=nan_weights)
+
+    def test_sample_none_weights(self, obj):
+        # Check None are also replaced by zeros.
+        weights_with_None = [None] * 10
+        weights_with_None[5] = 0.5
+        tm.assert_equal(
+            obj.sample(n=1, axis=0, weights=weights_with_None), obj.iloc[5:6]
+        )
+
+    @pytest.mark.parametrize(
+        "func_str,arg",
+        [
+            ("np.array", [2, 3, 1, 0]),
+            pytest.param(
+                "np.random.MT19937",
+                3,
+                marks=pytest.mark.skipif(np_version_under1p17, reason="NumPy<1.17"),
+            ),
+            pytest.param(
+                "np.random.PCG64",
+                11,
+                marks=pytest.mark.skipif(np_version_under1p17, reason="NumPy<1.17"),
+            ),
+        ],
+    )
+    @pytest.mark.parametrize("klass", [Series, DataFrame])
+    def test_sample_random_state(self, func_str, arg, klass):
+        # GH#32503
+        obj = DataFrame({"col1": range(10, 20), "col2": range(20, 30)})
+        if klass is Series:
+            obj = obj["col1"]
```

Comment:
Would it be better to have ``obj`` explicitly parametrized without performing checks inside the test method?

```
@pytest.mark.parametrize("obj", [Series(...), DataFrame({...})])
```

Reply:
i find this cleaner, also smaller memory footprint because constructing outside gets done at test collection time instead of test runtime

Reply:
OK, understood. Thank you for clarification on the footprint!